Mathlib/Algebra/FreeMonoid/Basic.lean

```diff
@@ -254,4 +377,68 @@ instance uniqueUnits : Unique (FreeMonoid α)ˣ where
     have : toList u.val ++ toList u.inv = [] := DFunLike.congr_arg toList u.val_inv
     (List.append_eq_nil.mp this).1
 
+@[to_additive]
+theorem map_surjective {f : α → β} : Function.Surjective f → Function.Surjective (map f) := by
+  intro fs d
+  induction' d using FreeMonoid.inductionOn' with head tail ih
+  · use 1
+    rfl
+  specialize fs head
+  rcases fs with ⟨a, rfl⟩
+  rcases ih with ⟨b, rfl⟩
+  use FreeMonoid.of a * b
+  rfl
+
+end Map
+
+/-! ### reverse -/
+
+section Reverse
+/-- reverses the symbols in a free monoid element -/
+@[to_additive "reverses the symbols in an additive free monoid element"]
+def reverse : FreeMonoid α → FreeMonoid α := List.reverse
+
+@[to_additive (attr := simp)]
+theorem reverse_of (a : α) : reverse (of a) = of a := rfl
+
+@[to_additive]
+theorem reverse_mul {a b : FreeMonoid α} : reverse (a * b) = reverse b * reverse a :=
+  List.reverse_append _ _
+
+@[to_additive (attr := simp)]
+theorem reverse_reverse {a : FreeMonoid α} : reverse (reverse a) = a := by
+  apply List.reverse_reverse
+
+@[to_additive (attr := simp)]
+theorem length_reverse {a : FreeMonoid α} : a.reverse.length = a.length :=
+  List.length_reverse _
+
+end Reverse
+
+/-- if two types are isomorphic, the free monoids over those types are isomorphic -/
+@[to_additive "if two types are isomorphic, the additive free monoids over those types are
+isomorphic"]
+def congr_iso {α : Type u_1} {β : Type u_2} (e : α ≃ β) : FreeMonoid α ≃* FreeMonoid β := by
+  apply MulEquiv.mk' ⟨FreeMonoid.map e.toFun, FreeMonoid.map e.invFun, _, _⟩
+  · simp
+  all_goals
+  intro x
+  simp [map_map]
+
+/-- given an isomorphism between α and β, convert a relation predicate to
+have an underlying type of β -/
+@[to_additive "given an isomorphism between α and β, convert a relation predicate to
+have an underlying type of β"]
+def mapRel (e : α ≃ β) (rel : FreeMonoid α → FreeMonoid α → Prop) :
+    FreeMonoid β → FreeMonoid β  → Prop :=
+  fun a b ↦ rel (FreeMonoid.congr_iso e.symm a) (FreeMonoid.congr_iso e.symm b)
+
+/-- given an isomorphism between α and β, pull back a relation predicate with underlying type β to
+one with underlying type α -/
+@[to_additive "given an isomorphism between α and β, pull back a relation predicate with underlying
+type β to one with underlying type α "]
+def comapRel (e : α ≃ β) (rel : FreeMonoid β → FreeMonoid β → Prop) :
+    FreeMonoid α → FreeMonoid α → Prop :=
+  fun a b ↦ rel (FreeMonoid.congr_iso e a) (FreeMonoid.congr_iso e b)
```

Comment:
Can you move these three to a later PR? The rest of this PR is ready to be merged, but these three feel less mature and I want comments from other people

Reply:
sure, it's now removed from this PR